src/topology/config/watcher.rs

```diff
@@ -0,0 +1,113 @@
+use super::Config;
+use futures::{Async, Future, Stream};
+use notify::{watcher, DebouncedEvent, RecursiveMode, Watcher};
+use std::{
+    path::Path,
+    sync::mpsc::{channel, TryRecvError},
+    time::{Duration, Instant},
+};
+use tokio::timer::Delay;
+
+// Per notify own documentation, it's advised to have delay of more than 30 sec,
+// so to avoid receiving repetitions of previous events on macOS.
+// Larger delays hurt responsivity, but that's fine as this is primarily designed
+// for cloud usage which isn't exactly super responsive.
+const CONFIG_WATCH_DELAY: std::time::Duration = std::time::Duration::from_secs(31);
+
+/// Returns when config should be reloaded.
+/// Never errors, and never ends.
+pub fn config_watcher(
+    config: &Config,
+    config_path: &Path,
+) -> crate::Result<impl Stream<Item = (), Error = ()>> {
+    config_watcher_with(config, config_path, CONFIG_WATCH_DELAY)
+}
+
+fn config_watcher_with(
+    config: &Config,
+    config_path: &Path,
+    delay: Duration,
+) -> crate::Result<impl Stream<Item = (), Error = ()>> {
+    let recv_delay = delay.checked_div(2).unwrap();
+
+    // Construct notify watcher
+    let mut stream = if config.global.reload_config {
+        info!("Creating configuration file watcher");
+        let (sender, receiver) = channel();
+        let mut watcher = watcher(sender, delay)?;
+        watcher.watch(config_path, RecursiveMode::NonRecursive)?;
+
+        Some((watcher, receiver, Delay::new(Instant::now() + recv_delay)))
+    } else {
+        None
+    };
+
+    // Turn mpsc::Receiver into a Stream
+    Ok(futures::stream::poll_fn(move || {
+        if let Some((_, recv, delay)) = stream.as_mut() {
+            // Check for events
+            loop {
+                match recv.try_recv() {
```

Comment:
we should use a `futures::sync::mpsc` channel here instead of using a regular sync channel. You can use this https://docs.rs/futures/0.1.29/futures/sync/mpsc/struct.Sender.html#method.try_send off task and in a sync fn to get data into an async notification context.

Reply:
I am not sure I understood you.

`notify 4.0.14` watcher requires `Sender` from `std`, so channel from `futures` can't be used.

Or, did you have something else in mind?

Reply:
In this case I think it might make sense to actually spawn a thread and do a `recv` then use a futures channel to trigger the stream, this way we won't show weird behavior with waiting and busy looping. I also am not the biggest fan of using the `poll_fn` so if we could break that out that would be nice as well.

Reply:
You are right, that would be nicer.